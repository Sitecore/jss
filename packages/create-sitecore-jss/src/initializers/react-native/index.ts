import path from 'path';
import { prompt } from 'inquirer';
import { prompts, ReactNativeAnswer } from './prompts';
import { Initializer } from '../../common/Initializer';
import { transform } from '../../common/steps';
import { ReactNativeArgs } from './args';

export default class ReactNativeInitializer implements Initializer {
  get isBase() {
    return true;
  }

  async init(args: ReactNativeArgs) {
    const defaults = args.yes
      ? {
          appName: 'sitecore-jss-react-native',
          hostName: 'sitecore-jss-react-native.dev.local',
          language: '',
        }
      : {};

    const answers = await prompt<ReactNativeAnswer>(prompts, { ...defaults, ...args });

    const mergedArgs = {
      ...args,
      ...answers,
    };

    const templatePath = path.resolve(__dirname, '../../templates/react-native');
    await transform(templatePath, mergedArgs, {
      filter: (filePath: string) =>
        !!mergedArgs.language || !filePath.endsWith('{{language}}.json'),
    });

    const response = {
      appName: answers.appName,
    };

    return response;
  }
}
