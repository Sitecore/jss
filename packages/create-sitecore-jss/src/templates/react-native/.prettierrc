{
  "endOfLine": "crlf",
  "singleQuote": true,
  "trailingComma": "es5",
  "tabWidth": 2,
  "arrowParens": "always",
  "printWidth": 100
}
