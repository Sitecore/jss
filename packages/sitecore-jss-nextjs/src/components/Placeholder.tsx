import React, { useContext } from 'react';
import {
  Placeholder as ReactPlaceholder,
  PlaceholderComponentProps,
} from '@sitecore-jss/sitecore-jss-react';
import { ComponentPropsReactContext } from './ComponentPropsContext';
import { ComponentProps } from '@sitecore-jss/sitecore-jss-react/types/components/PlaceholderCommon';

export const Placeholder = (props: PlaceholderComponentProps) => {
  const componentPropsContext = useContext(ComponentPropsReactContext);

  function getContextProps(initialProps: ComponentProps) {
    if (!initialProps.rendering.uid) return initialProps;

    const data = componentPropsContext[initialProps.rendering.uid] as {
      [key: string]: unknown;
    };

    return data;
  }

  return (
    <ReactPlaceholder
      {...props}
      modifyComponentProps={(initialProps) => {
        const finalProps = props.modifyComponentProps
          ? props.modifyComponentProps(initialProps)
          : initialProps;
        return { ...finalProps, ...getContextProps(initialProps) };
      }}
    />
  );
};
