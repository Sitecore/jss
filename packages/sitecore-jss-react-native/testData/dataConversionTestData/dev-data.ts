export const devRouteData = {
  name: 'portfolio',
  displayName: 'Portfolio',
  fields: {
    metaTitle: {
      value: 'Portfolio',
    },
    titleText: {
      value: 'Portfolio',
    },
    body: {
      value: "You'll be amazed by all our good work.",
    },
  },
  placeholders: {
    main: [
      {
        componentName: 'Portfolio',
        fields: {},
        params: {},
        dataSource: '',
        placeholders: {
          'page-header': [
            {
              componentName: 'Jumbotron',
              dataSource: '',
              params: {
                shade: 'light',
                titleSize: '2',
              },
            },
          ],
          'page-content': [
            {
              componentName: 'Carousel',
              dataSource: '',
              fields: {
                items: [
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/1.jpg',
                        },
                      },
                      title: {
                        value: 'First slide label!',
                      },
                      body: {
                        value: '<p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>',
                      },
                    },
                  },
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/2.jpg',
                        },
                      },
                      title: {
                        value: 'Second slide label',
                      },
                      body: {
                        value: '<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>',
                      },
                    },
                  },
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/3.jpg',
                        },
                      },
                      title: {
                        value: 'Third slide label',
                      },
                      body: {
                        value:
                          '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>',
                      },
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ],
  },
};

export const convertedDevRouteData = {
  name: 'portfolio',
  displayName: 'Portfolio',
  fields: {
    metaTitle: {
      value: 'Portfolio',
      editable: 'Portfolio',
    },
    body: {
      value: "You'll be amazed by all our good work.",
      editable: "You'll be amazed by all our good work.",
    },
    titleText: {
      value: 'Portfolio',
      editable: 'Portfolio',
    },
  },
  placeholders: {
    main: [
      {
        componentName: 'Portfolio',
        dataSource: '',
        fields: {},
        params: {},
        placeholders: {
          'page-header': [
            {
              componentName: 'Jumbotron',
              dataSource: '',
              params: {
                shade: 'light',
                titleSize: '2',
              },
            },
          ],
          'page-content': [
            {
              componentName: 'Carousel',
              dataSource: '',
              fields: {
                items: [
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/1.jpg',
                        },
                        editable: '',
                      },
                      title: {
                        value: 'First slide label!',
                        editable: 'First slide label!',
                      },
                      body: {
                        value: '<p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>',
                        editable:
                          '<p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>',
                      },
                    },
                  },
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/2.jpg',
                        },
                        editable: '',
                      },
                      title: {
                        value: 'Second slide label',
                        editable: 'Second slide label',
                      },
                      body: {
                        value: '<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>',
                        editable: '<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>',
                      },
                    },
                  },
                  {
                    fields: {
                      image: {
                        value: {
                          src: '/assets/img/portfolio/3.jpg',
                        },
                        editable: '',
                      },
                      title: {
                        value: 'Third slide label',
                        editable: 'Third slide label',
                      },
                      body: {
                        value:
                          '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>',
                        editable:
                          '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>',
                      },
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ],
  },
};
