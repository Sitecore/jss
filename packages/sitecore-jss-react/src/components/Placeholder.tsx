import React from 'react';
import { PlaceholderCommon, PlaceholderProps } from './PlaceholderCommon';
import { withComponentFactory } from '../enhancers/withComponentFactory';
import { ComponentRendering, HtmlElementRendering } from '@sitecore-jss/sitecore-jss';

export interface PlaceholderComponentProps extends PlaceholderProps {
  /**
   * Render props function that enables control over the rendering of the components in the placeholder.
   * Useful for techniques like wrapping each child in a wrapper component.
   */
  render?: (components: React.ReactNode[], data: (ComponentRendering | HtmlElementRendering)[], props: PlaceholderProps) => React.ComponentClass<any> | React.SFC<any>;

  /**
   * Render props function that is called for each non-system component added to the placeholder.
   * Mutually exclusive with `render`. System components added during Experience Editor are automatically rendered as-is.
   */
  renderEach?: (components: React.ReactNode[], data: (ComponentRendering | HtmlElementRendering)[], props: PlaceholderProps) => React.ComponentClass<any> | React.SFC<any>;
}

function isRawRendering(rendering: HtmlElementRendering | ComponentRendering): rendering is HtmlElementRendering {
  return !(rendering as ComponentRendering).componentName && (rendering as HtmlElementRendering).name !== undefined;
}

class PlaceholderComponent extends PlaceholderCommon {
  static propTypes = PlaceholderCommon.propTypes;

  constructor(props: PlaceholderComponentProps) {
    super(props);
  }

  render() {
    let childProps: any = { ...this.props };

    delete childProps.componentFactory;

    if (this.state.error) {
      if (childProps.errorComponent) {
        return <childProps.errorComponent error={this.state.error} />;
      }

      return (
        <div className="sc-jss-placeholder-error">
          A rendering error occurred: {this.state.error.message}.
        </div>
      );
    }

    const renderingData = childProps.rendering;

    const placeholderData = PlaceholderCommon.getPlaceholderDataFromRenderingData(renderingData, this.props.name);
    const components =  this.getComponentsForRenderingData(placeholderData);

    if (this.props.renderEmpty && placeholderData.every((rendering: ComponentRendering | HtmlElementRendering) => isRawRendering(rendering))) {
      return this.props.renderEmpty(components, placeholderData, childProps);
    } else if (this.props.render) {
      return this.props.render(components, placeholderData, childProps);
    } else if (this.props.renderEach) {
      return components.map((component, index) => {
        if (component && component.props && component.props.type === 'text/sitecore') {
          return component;
        }

        return this.props.renderEach(component, index);
      });
    } else {
      return components;
    }
  }
}

export const Placeholder = withComponentFactory(PlaceholderComponent);