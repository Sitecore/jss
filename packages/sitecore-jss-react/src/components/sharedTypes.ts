import { ComponentType } from 'react';

export type ComponentFactory = (componentName: string) => ComponentType | null;
