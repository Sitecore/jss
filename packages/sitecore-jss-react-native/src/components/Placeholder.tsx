import React from 'react';
import { View } from 'react-native';
import { withComponentFactory } from '../enhancers/withComponentFactory';
import { ComponentRendering, HtmlElementRendering } from '@sitecore-jss/sitecore-jss';
import { PlaceholderCommon, PlaceholderProps } from './PlaceholderCommon';

export interface PlaceholderComponentProps extends PlaceholderProps {
  /**
   * Render props function that enables control over the rendering of the components in the placeholder.
   * Useful for techniques like wrapping each child in a wrapper component.
   */
  render?: (
    components: React.ReactNode[],
    data: (ComponentRendering | HtmlElementRendering)[],
    props: PlaceholderProps
  ) => React.ComponentClass<unknown> | React.SFC<unknown>;

  /**
   * Render props function that is called when the placeholder contains no content components.
   * Can be used to wrap the Sitecore EE empty placeholder markup in something that's visually correct
   */
  renderEmpty?: (
    components: React.ReactNode[],
    data: (ComponentRendering | HtmlElementRendering)[],
    props: PlaceholderProps
  ) => React.ComponentClass<unknown> | React.SFC<unknown> | React.ReactNode;

  /**
   * Render props function that is called for each non-system component added to the placeholder.
   * Mutually exclusive with `render`. System components added during Experience Editor are automatically rendered as-is.
   */
  renderEach?: (
    component: React.ReactNode,
    index: number
  ) => React.ComponentClass<unknown> | React.SFC<unknown> | React.ReactNode;
}

/**
 * @param {HtmlElementRendering | ComponentRendering} rendering
 */
function isRawRendering(
  rendering: HtmlElementRendering | ComponentRendering
): rendering is HtmlElementRendering {
  return (
    !(rendering as ComponentRendering).componentName &&
    (rendering as HtmlElementRendering).name !== undefined
  );
}

class PlaceholderComponent extends PlaceholderCommon {
  static propTypes = PlaceholderCommon.propTypes;

  constructor(props: PlaceholderComponentProps) {
    super(props);
  }

  render() {
    const props: PlaceholderComponentProps = this.props;
    const childProps: PlaceholderComponentProps = { ...this.props };

    delete childProps.componentFactory;
    delete childProps.render;
    delete childProps.renderEach;

    if (this.state.error) {
      return <View>A rendering error occurred: {this.state.error.message}.</View>;
    }

    const renderingData = childProps.rendering;

    const placeholderData = PlaceholderCommon.getPlaceholderDataFromRenderingData(
      renderingData,
      props.name
    );
    const components = this.getComponentsForRenderingData(placeholderData);

    if (
      props.renderEmpty &&
      placeholderData.every((rendering: ComponentRendering | HtmlElementRendering) =>
        isRawRendering(rendering)
      )
    ) {
      return components && props.renderEmpty(components, placeholderData, childProps);
    } else if (props.render) {
      return components && props.render(components, placeholderData, childProps);
    } else if (props.renderEach) {
      return (
        components &&
        components.map((component, index) => {
          if (component && component.props && component.props.type === 'text/sitecore') {
            return component;
          }

          return props.renderEach && props.renderEach(component, index);
        })
      );
    } else {
      return components;
    }
  }
}

export const Placeholder = withComponentFactory(PlaceholderComponent);
