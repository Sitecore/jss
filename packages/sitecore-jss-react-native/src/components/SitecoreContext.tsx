import React from 'react';
import PropTypes from 'prop-types';
import { ComponentFactory } from './sharedTypes';

export interface SitecoreContextProps {
  componentFactory: ComponentFactory;
  children: unknown;
}

export const ComponentFactoryReactContext = React.createContext<ComponentFactory>(
  {} as ComponentFactory
);

export class SitecoreContext extends React.Component<SitecoreContextProps> {
  static propTypes = {
    children: PropTypes.any.isRequired,
    componentFactory: PropTypes.func,
  };

  static displayName = 'SitecoreContext';

  componentFactory: ComponentFactory;

  constructor(props: SitecoreContextProps, context: unknown) {
    super(props, context);
    this.componentFactory = props.componentFactory;
  }

  render() {
    return (
      <ComponentFactoryReactContext.Provider value={this.componentFactory}>
        {this.props.children}
      </ComponentFactoryReactContext.Provider>
    );
  }
}
