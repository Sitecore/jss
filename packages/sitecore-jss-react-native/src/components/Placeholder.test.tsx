import { View } from 'react-native';
import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import { ComponentRendering, RouteData } from '@sitecore-jss/sitecore-jss';
import { Placeholder, PlaceholderComponentProps } from './Placeholder';
import { SitecoreContext } from './SitecoreContext';
import { devData } from '../../testData/dev-data';
import { lsDataEeOff } from '../../testData/LS-data-EE-off';
import {ComponentFactory} from './sharedTypes';

const componentFactory = (componentName: string) => {
  const components = new Map();

  // pass otherProps to page-content to test property cascading through the Placeholder
  const Home: React.SFC<PlaceholderComponentProps> = ({ rendering, ...otherProps }) => (
    <View testID="home-mock">
      <Placeholder name="page-header" rendering={rendering} />
      <Placeholder name="page-content" rendering={rendering} {...otherProps} />
    </View>
  );
  Home.propTypes = {
    placeholders: PropTypes.object,
    rendering: PropTypes.any,
  };

  components.set('Home', Home);

  type DownloadCalloutProps = {
    fields: {
      message?: {
        value: string;
      }
    }
  }

  const DownloadCallout: React.SFC<DownloadCalloutProps> = (props) => (
    <View testID="download-callout-mock">
      {props.fields.message ? props.fields.message.value : null}
    </View>
  );
  DownloadCallout.propTypes = {
    fields: PropTypes.any
  };

  components.set('DownloadCallout', DownloadCallout);
  components.set('Jumbotron', () => <View testID="jumbotron-mock" />);

  return components.get(componentName);
};

describe('<Placeholder />', () => {
  const testData = [
    { label: 'Dev data', data: devData },
    { label: 'LayoutService data - EE off', data: lsDataEeOff },
  ];

  testData.forEach((dataSet) => {
    describe(`with ${dataSet.label}`, () => {
      test('should render a placeholder with given key', () => {
        const phData = dataSet.data.sitecore.route.placeholders.main as unknown as ComponentRendering[];
        const component = phData.find((c: ComponentRendering) => c.componentName);
        const phKey = 'page-content';

        // this will create a snapshot of the rendered component.
        // you'll need to visually inspect the snapshot to ensure expected rendering.
        const renderedComponent = renderer
          .create(
            <Placeholder name={phKey} rendering={component as ComponentRendering} componentFactory={componentFactory} />
          )
          .toJSON();

        // subsequent test runs will then compare the rendered output to the snapshot.
        expect(renderedComponent).toMatchSnapshot();
      });

      test('should render nested placeholders and props correctly', () => {
        const component = dataSet.data.sitecore.route as unknown as RouteData;
        const phKey = 'main';

        const renderedComponent = renderer
          .create(
            <SitecoreContext componentFactory={componentFactory}>
              <Placeholder name={phKey} rendering={component} />
            </SitecoreContext>
          )
          .toJSON();

        expect(renderedComponent).toMatchSnapshot();
      });
    });
  });

  it('should render missing component for unresolved/unknown components', () => {
    const errorSpy = jest.spyOn(console, 'error');
    errorSpy.mockImplementation(() => {});

    // use local mocks to avoid console warnings from Placeholder about unknown nested placeholders
    const Home = () => <View testID="home-mock" />;
    const factory: ComponentFactory = (componentName: string) => (componentName === 'Home' ? Home : null);

    const route = {
      placeholders: {
        main: [
          {
            componentName: 'Home',
          },
          {
            componentName: 'whatisthis',
          },
        ],
      },
    };
    const phKey = 'main';

    const renderedComponent = renderer
      .create(<Placeholder name={phKey} rendering={route as unknown as RouteData} componentFactory={factory} />)
      .toJSON();

    // placeholder should only render "Home" component
    expect(renderedComponent).toMatchSnapshot();

    errorSpy.mockRestore();
  });

  it('should render null for unknown placeholder', () => {
    const warnSpy = jest.spyOn(console, 'warn');
    warnSpy.mockImplementation(() => {});
    const route: RouteData = {
      name: '',
      placeholders: {
        main: [
          {
            componentName: 'Home',
          },
        ],
      },
    };
    const phKey = 'unknown';

    const renderedComponent = renderer
      .create(<Placeholder name={phKey} rendering={route} componentFactory={componentFactory} />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
    expect(warnSpy).toHaveBeenCalledTimes(1);
    warnSpy.mockRestore();
  });
});
