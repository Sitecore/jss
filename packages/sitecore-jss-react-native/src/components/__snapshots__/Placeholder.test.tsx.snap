// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Placeholder /> should render missing component for unresolved/unknown components 1`] = `
Array [
  <View
    testID="home-mock"
  />,
  <View
    style={
      Object {
        "backgroundColor": "darkorange",
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginVertical": 20,
        }
      }
    >
      whatisthis
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "marginVertical": 20,
        }
      }
    >
      JSS component is missing React Native implementation.
    </Text>
  </View>,
]
`;

exports[`<Placeholder /> should render null for unknown placeholder 1`] = `null`;

exports[`<Placeholder /> with Dev data should render a placeholder with given key 1`] = `
<View
  testID="download-callout-mock"
/>
`;

exports[`<Placeholder /> with Dev data should render nested placeholders and props correctly 1`] = `
<View
  testID="home-mock"
>
  <View
    testID="jumbotron-mock"
  />
  <View
    testID="download-callout-mock"
  >
    JavaScript all the things!
  </View>
</View>
`;

exports[`<Placeholder /> with LayoutService data - EE off should render a placeholder with given key 1`] = `
<View
  testID="download-callout-mock"
/>
`;

exports[`<Placeholder /> with LayoutService data - EE off should render nested placeholders and props correctly 1`] = `
<View
  testID="home-mock"
>
  <View
    testID="jumbotron-mock"
  />
  <View
    testID="download-callout-mock"
  >
    JavaScript all the things!
  </View>
</View>
`;
